src/vos/vos_io.c

```diff
@@ -1532,9 +1528,13 @@ vos_fetch_begin(daos_handle_t coh, daos_unit_oid_t oid, daos_epoch_t epoch,
 	if (rc != 0) {
 		daos_recx_ep_list_free(ioc->ic_recx_lists, ioc->ic_iod_nr);
 		ioc->ic_recx_lists = NULL;
-		return vos_fetch_end(vos_ioc2ioh(ioc), NULL, rc);
+		rc = vos_fetch_end(vos_ioc2ioh(ioc), NULL, rc);
+		*ioh = DAOS_HDL_INVAL;
```

Comment:
Generally, as you pointed out, the caller needs to check the return value before accessing the output "ioh". That is the regular logic. But sometime, be as some cleanup logic, the caller may check the "ioh" even if under "rc != 0" case, because such "rc != 0" may be caused by other reason, such as "if (rc != 0) goto out;" and check "ioh" behind "out" label.
 
Here, in my patch, I set DAOS_HDL_INVAL under failure case just for such purpose. So if need refresh the patch, we should set "ioh" as DAOS_HDL_INVAL under all failure cases in vos_fetch_begin(). I think we can do that in another PR since it is not related with this ticket. Then we will not block 2.4 landing. How do you think?